mart/attack/perturber/perturber.py

```diff
@@ -78,29 +89,51 @@ def initialize_parameters(self, input, target):
 
         self.initializer(self.perturbation)
 
-    def parameter_groups(self):
-        """Return parameters along with the pre-defined optimization parameters.
+        # Initialize optimizer
+        self.opt = self.optimizer_fn([self.perturbation])
 
-        Example: `[{"params": perturbation, "lr":0.1, "momentum": 0.9}]`
-        """
-        if "params" in self.optim_params:
-            raise ValueError(
-                'Optimization parameters should not include "params" which will override the actual parameters to be optimized. '
-            )
+    def on_advance_start(
+        self,
+        *,
+        adversary: Adversary,
+        input: torch.Tensor | tuple,
+        target: torch.Tensor | dict[str, Any] | tuple,
+        model: torch.nn.Module,
+        **kwargs,
+    ):
+        self.opt.zero_grad()
 
-        return [{"params": self.perturbation} | self.optim_params]
+    def on_advance_end(
+        self,
+        *,
+        adversary: Adversary,
+        input: torch.Tensor | tuple,
+        target: torch.Tensor | dict[str, Any] | tuple,
+        model: torch.nn.Module,
+        **kwargs,
+    ):
+        self.opt.step()
```

Comment:
We may have to rename `advance` then, because `advance` means `self.opt.step()`.

Reply:
Yeah...I think we should really use the existing loop interfaces in PL. I'll look into making Perturber a proper optimizer.

Reply:
The current implementation is fine because a perturber always executes before the